File: GitToolBox/src/main/java/zielu/gittoolbox/fetch/AutoFetchAllowedBuild.java
package zielu.gittoolbox.fetch;

import com.intellij.compiler.server.BuildManagerListener;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import com.intellij.util.messages.MessageBusConnection;
import java.util.Objects;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicBoolean;
import zielu.gittoolbox.extension.AutoFetchAllowed;

public class AutoFetchAllowedBuild implements AutoFetchAllowed {
  private final Logger log = Logger.getInstance(getClass());

  private final AtomicBoolean buildRunning = new AtomicBoolean();
  private MessageBusConnection connection;

  @Override
  public boolean isAllowed() {
    return !buildRunning.get();
  }

  @Override
  public void initialize(Project project) {
    connection = project.getMessageBus().connect();
    connection.subscribe(BuildManagerListener.TOPIC, new BuildManagerListener() {
      @Override
      public void beforeBuildProcessStarted(Project project, UUID sessionId) {
        //do nothing
      }

      @Override
      public void buildStarted(Project currentProject, UUID sessionId, boolean isAutomake) {
        log.debug("Build started");
        if (Objects.equals(project, currentProject)) {
          buildRunning.set(true);
        }
      }

      @Override
      public void buildFinished(Project currentProject, UUID sessionId, boolean isAutomake) {
        log.debug("Build finished");
        if (Objects.equals(project, currentProject)) {
          buildRunning.set(false);
          fireStateChanged(project);
        }
      }
    });
  }

  @Override
  public void dispose() {
    if (connection != null) {
      connection.disconnect();
      connection = null;
    }
  }
}

File: GitToolBox/src/main/kotlin/zielu/gittoolbox/fetch/AutoFetchAllowedBuild.kt
package zielu.gittoolbox.fetch

import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.Project
import zielu.gittoolbox.util.AppUtil
import java.util.Optional
import java.util.concurrent.atomic.AtomicBoolean

internal class AutoFetchAllowedBuild(private val project: Project) {
  private val gateway = AutoFetchAllowedLocalGateway(project)
  private val buildRunning = AtomicBoolean()

  fun isFetchAllowed(): Boolean {
    return !buildRunning.get()
  }

  fun onBuildStarted(builtProject: Project) {
    log.debug("Build started")
    if (isCurrentProject(builtProject)) {
      buildRunning.set(true)
    }
  }

  fun onBuildFinished(builtProject: Project) {
    log.debug("Build finished")
    if (isCurrentProject(builtProject)) {
      if (buildRunning.compareAndSet(true, false)) {
        gateway.fireStateChanged()
      }
    }
  }

  private fun isCurrentProject(builtProject: Project): Boolean {
    return project == builtProject
  }

  companion object {
    private val log = Logger.getInstance(AutoFetchAllowedBuild::class.java)

    fun getInstance(project: Project): Optional<AutoFetchAllowedBuild> {
      return AppUtil.getServiceInstanceSafe(project, AutoFetchAllowedBuild::class.java)
    }
  }
}
